src/main/java/io/numaproj/pulsar/consumer/PulsarSource.java

```
@@ -7,139 +7,119 @@
-    private final Map<String, org.apache.pulsar.client.api.Message<byte[]>> messages = new ConcurrentHashMap<>();
+    // Map tracking received messages (keyed by Pulsar message ID string)
+    private final Map<String, org.apache.pulsar.client.api.Message<byte[]>> messagesToAck = new ConcurrentHashMap<>();
 
     private Server server;
 
     @Autowired
-    private PulsarClient pulsarClient;
-
-    @Autowired
-    private Consumer<byte[]> pulsarConsumer;
+    private PulsarConsumerManager pulsarConsumerManager;
 
     @PostConstruct
     public void startServer() throws Exception {
-
         server = new Server(this);
         server.start();
         server.awaitTermination();
     }
 
     @Override
     public void read(ReadRequest request, OutputObserver observer) {
-        long startTime = System.currentTimeMillis();
-
-        // minimal check: if outstanding messages haven't been acknowledged, do nothing.
-        if (!messages.isEmpty()) {
+        // If there are messages not acknowledged, return
+        if (!messagesToAck.isEmpty()) {
             return;
         }
 
-        for (int i = 0; i < request.getCount(); i++) {
-            if (System.currentTimeMillis() - startTime > request.getTimeout().toMillis()) {
-                return;
-            }
+        Consumer<byte[]> consumer = null;
 
+        try {
+            // Obtain a consumer with the desired settings.
+            consumer = pulsarConsumerManager.getOrCreateConsumer(request.getCount(), request.getTimeout().toMillis());
+            Messages<byte[]> batchMessages = null;
             try {
-                org.apache.pulsar.client.api.Message<byte[]> pMsg = pulsarConsumer
-                        .receive((int) request.getTimeout().toMillis(), TimeUnit.MILLISECONDS);
-                if (pMsg == null) {
-                    return;
-                }
+                batchMessages = consumer.batchReceive();
+            } catch (PulsarClientException.AlreadyClosedException ace) {
```

Ideally we should never get this exception. If so, it means something went wrong with the system. And in that case, I would prefer to throw the exception and stop consuming as opposed to trying to re-create a new connection.

Let's throw any Exception and monitor the behaviour during load testing and see if we actually see `AlreadyClosedException`.